modules/library/xml/src/main/java/org/geotools/xml/resolver/SchemaCache.java

```diff
@@ -219,48 +217,38 @@ static byte[] download(URI location) {
         return download(location, DEFAULT_DOWNLOAD_BLOCK_SIZE);
     }
 
-    static byte[] download(URI location, int blockSize) {
-        return download(location, blockSize, 0);
-    }
-
     /**
      * Retrieve the contents of a remote URL.
      *
      * @param location and absolute http/https URL.
      * @param blockSize download block size
-     * @param redirectionCount the number of redirection attempts already performed while
-     *     downloading
      * @return the bytes contained by the resource, or null if it could not be downloaded
      */
-    static byte[] download(URI location, int blockSize, int redirectionCount) {
+    static byte[] download(URI location, int blockSize) {
         try {
             URL url = location.toURL();
             String protocol = url.getProtocol();
             if (protocol == null || !(protocol.equals("http") || protocol.equals("https"))) {
                 LOGGER.warning("Unexpected download URL protocol: " + protocol);
                 return null;
             }
-            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
-            connection.setConnectTimeout(downloadTimeout);
-            connection.setReadTimeout(downloadTimeout);
-            connection.setUseCaches(false);
-            connection.connect();
-            int responseCode = connection.getResponseCode();
-            if (responseCode != HttpURLConnection.HTTP_OK) {
-                if (hasRedirect(responseCode))
-                    return followRedirect(connection, blockSize, redirectionCount);
-                LOGGER.warning(
-                        String.format(
-                                "Unexpected response \"%d %s\" while downloading %s",
-                                connection.getResponseCode(),
-                                connection.getResponseMessage(),
-                                location.toString()));
+            HTTPClient httpClient = HTTPClientFinder.createClient();
```

Comment:
Interesting bit here... it switches from HttpURLConnection to something that may, or may not, be a SimpleHttpClient (depends on the default system hints AFAIK).

I'm not sure what would happen if the multithreaded client (commons-http base) is used instead. Does it have a redirection limit? In GeoServer, when we want to ensure a simple HTTP client is used, we do the following:

```java
client =
                    HTTPClientFinder.createClient(new Hints(Hints.HTTP_CLIENT, SimpleHttpClient.class));
```

However, at the same time, doing so would break the tests below, which rely on default system hints to inject a mock client. Hum... maybe just verify that if the multithreaded client is used, we do not risk a loop of redirects?

Reply:
Good point, I did some testing and found out the Apache client does check for circular redirections and raises an IOException in this case (in this class: https://www.javadoc.io/doc/org.apache.httpcomponents/httpclient/4.3.5/org/apache/http/impl/execchain/RedirectExec.html, default is 50 redirects). With the current code the exception is caught and logged by the schema cache (similarly to >= 400 status codes).
I added a test for this, let me know what you think. Thanks in advance